### DISCUSSION TITLE
DummyNanoClock vs SystemNanoClock?!

### DISCUSSION THREAD

**Discussion by javafanboy** [apache/logging-log4j2#3197] [Q&A] [answered]
I am trying to understand how the various clock implementations in log4j2 works and I found the org.apache.logging.log4j.core.util.SystemNanoClock in the documentation but for me another class DummyNanoClock is instead used (or at least that is what is returned when I call the getNanoClock() method in LogManager.getContext(false).getConfiguration()). What is the use of these two classes? Why is the "dummy" used and when /why is SystemNanoClock used?

There is also a PreciseClock interface but this also seem to only return time with millisecond precision... 

I have specified `PatternLayout pattern="%d{yyyy-MM-dd HH:mm:ss.nnnnnnnnn} [%t] %-5level %logger{36} - %msg%n">`for my appender and the "nnnnnnnnn" field is actually populated but only with micro second precision (perhaps this is the best MacOS on M3 ARM provides) but how is this possible if the "dummy" nano clock is used?

Just for fun I tried specifying the property -Dlog4j.Clock=org.apache.logging.log4j.core.util.SystemNanoClock but this resulted in a class cast exception so this setting seem to be for setting some other clock and not the "nanoclock" or what am I missing here...?!

Anybody that can explain?

**Comment by ppkarwasz** [comment, marked as answer]
In Log4j Core there two **different** types of time sources:

- [`o.a.l.l.core.util.Clock`](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/util/Clock.html) and its subinterface [`o.a.l.l.core.util.PreciseClock`](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/time/PreciseClock.html): these sources return the current date and time with different degrees of precision. On POSIX systems they correspond to the [`CLOCK_REALTIME` constant of `clock_gettime`](https://man7.org/linux/man-pages/man3/clock_gettime.3.html).
This is the source used by the [date pattern converter (`%d`)](https://logging.apache.org/log4j/2.x/manual/pattern-layout.html#converter-date).
- [`o.a.l.l.core.util.NanoClock`](https://logging.apache.org/log4j/2.x/javadoc/log4j-core/org/apache/logging/log4j/core/util/NanoClock.html) returns the numbers of nanoseconds from an **arbitrary** point in time. On POSIX systems they correspond to the `CLOCK_MONOTONIC` constant. On Linux this gives the number of nanoseconds from the **system boot time**.
This is the source used by the [nanosecond pattern converter (`%N`)](https://logging.apache.org/log4j/2.x/manual/pattern-layout.html#converter-nano).

Since you are using `%d`, you want a `PreciseClock` implementation, not a `NanoClock` implementation.
Currently Log4j Core contains a single implementation of that interface: the version of [`o.a.l.l.core.util.SystemClock`](https://github.com/apache/logging-log4j2/blob/2.x/log4j-core-java9/src/main/java/org/apache/logging/log4j/core/util/SystemClock.java) for Java 9 or later.
In order to use it you need to:

- Run on JRE 9 or later.
- Make sure you didn't break the **multi-release** `log4j-core` artifact. Using shading/shadowing is the most common way to break multi-release JARs. See [Shading/shadowing FAQ entry](https://logging.apache.org/log4j/2.x/faq.html#shading) for more information.


**Remark**: Note that in Log4j Core `2.25.0` the meaning of formatting patterns will be **change** to the one specified by [`DateTimeFormatter`](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html) and
 the recommended pattern for nanos will be `%d{yyyy-MM-dd HH:mm:ss.SSSSSSSSS}`.

**Comment by javafanboy** [comment]
Thanks for the good explanation - now more clear for me!
